src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt

```diff
@@ -42,16 +36,25 @@ class SymbolTableVisitor {
     fun getCurrentScopeID(): Int = currentScope.id
 }
 
-open class SymbolTableNode(val parent: SymbolTable, override val id: Int) : SymbolTable(id) {
+sealed class SymbolTable(open val id: Int) {
+    protected val table: MutableMap<String, IdentifierData> = mutableMapOf()
 
-    override fun getTypeOf(identifier: String): Type {
-        return table[identifier]?.type ?: parent.getTypeOf(identifier)
+    abstract operator fun get(identifier: String): IdentifierData
+
+    operator fun set(identifier: String, data: IdentifierData) {
+        table[identifier] = data
+    }
+}
+
+open class SymbolTableNode(val parent: SymbolTable, override val id: Int) : SymbolTable(id) {
+    override operator fun get(identifier: String): IdentifierData {
+        return table[identifier] ?: parent[identifier]
```

Comment:
just curious why don't we need to recurse here?

Reply:
so operator fun get is the array like access. So when we are doing parent[identifier] this is equivalent to parent.get(identifier)